File: cane-csharp/cane-core/Runtime/Geometry/BoundsMath.cs
using System;
using System.Collections.Generic;

namespace Cane.Geometry
{
    internal static class BoundsMath
    {
        private static bool OnSegment(double x, double y, double ax, double ay, double bx, double by)
        {
            double scale = Math.Max(1, Math.Max(Math.Max(Math.Abs(x), Math.Abs(y)), Math.Max(Math.Max(Math.Abs(ax), Math.Abs(ay)), Math.Max(Math.Abs(bx), Math.Abs(by)))));
            if (Math.Abs((x - ax) * (by - ay) - (y - ay) * (bx - ax)) > 2.2204460492503131e-16 * 32 * scale * scale) return false;
            return x >= Math.Min(ax, bx) && x <= Math.Max(ax, bx) && y >= Math.Min(ay, by) && y <= Math.Max(ay, by);
        }
        internal static bool Contains(IReadOnlyList<float> vertices, double x, double y)
        {
            if (vertices.Count < 6) return false; bool inside = false;
            for (int previous = vertices.Count - 2, current = 0; current < vertices.Count; previous = current, current += 2)
            {
                double ax = vertices[previous], ay = vertices[previous + 1], bx = vertices[current], by = vertices[current + 1];
                if (OnSegment(x, y, ax, ay, bx, by)) return true;
                if ((ay > y) != (by > y) && x < (ax - bx) * (y - by) / (ay - by) + bx) inside = !inside;
            }
            return inside;
        }
        private static double Orientation(double ax, double ay, double bx, double by, double cx, double cy) => (bx - ax) * (cy - ay) - (by - ay) * (cx - ax);
        private static bool Crosses(double ax, double ay, double bx, double by, double cx, double cy, double dx, double dy)
        {
            double a = Orientation(ax, ay, bx, by, cx, cy), b = Orientation(ax, ay, bx, by, dx, dy), c = Orientation(cx, cy, dx, dy, ax, ay), d = Orientation(cx, cy, dx, dy, bx, by);
            if ((a > 0 && b < 0 || a < 0 && b > 0) && (c > 0 && d < 0 || c < 0 && d > 0)) return true;
            return a == 0 && OnSegment(cx, cy, ax, ay, bx, by) || b == 0 && OnSegment(dx, dy, ax, ay, bx, by)
                || c == 0 && OnSegment(ax, ay, cx, cy, dx, dy) || d == 0 && OnSegment(bx, by, cx, cy, dx, dy);
        }
        internal static bool Segment(IReadOnlyList<float> v, double x1, double y1, double x2, double y2)
        {
            if (Contains(v, x1, y1) || Contains(v, x2, y2)) return true;
            for (int p = v.Count - 2, i = 0; i < v.Count; p = i, i += 2) if (Crosses(x1, y1, x2, y2, v[p], v[p + 1], v[i], v[i + 1])) return true; return false;
        }
        internal static bool Intersects(IReadOnlyList<float> a, IReadOnlyList<float> b)
        {
            if (a.Count == 0 || b.Count == 0) return false; if (Contains(a, b[0], b[1]) || Contains(b, a[0], a[1])) return true;
            for (int p = a.Count - 2, i = 0; i < a.Count; p = i, i += 2) if (Segment(b, a[p], a[p + 1], a[i], a[i + 1])) return true; return false;
        }
    }
}

File: cane-cpp/cane-core/src/geometry/bounds_math.cpp
#include "bounds_math.hpp"
#include <algorithm>
#include <cmath>
#include <limits>

namespace cane::geometry {
namespace {
struct DoublePoint { double x, y; };
DoublePoint point(const std::vector<float>& v, std::size_t i) noexcept { return {v[i], v[i + 1]}; }
bool on_segment(DoublePoint p, DoublePoint a, DoublePoint b) noexcept {
    const double scale = std::max({1.0, std::abs(p.x), std::abs(p.y), std::abs(a.x), std::abs(a.y), std::abs(b.x), std::abs(b.y)});
    const double cross = (p.x - a.x) * (b.y - a.y) - (p.y - a.y) * (b.x - a.x);
    if (std::abs(cross) > std::numeric_limits<double>::epsilon() * 32 * scale * scale) return false;
    return p.x >= std::min(a.x, b.x) && p.x <= std::max(a.x, b.x) && p.y >= std::min(a.y, b.y) && p.y <= std::max(a.y, b.y);
}
double orientation(DoublePoint a, DoublePoint b, DoublePoint c) noexcept {
    return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
}
bool crosses(DoublePoint a, DoublePoint b, DoublePoint c, DoublePoint d) noexcept {
    const double ac = orientation(a, b, c), ad = orientation(a, b, d), ca = orientation(c, d, a), cb = orientation(c, d, b);
    if (((ac > 0 && ad < 0) || (ac < 0 && ad > 0)) && ((ca > 0 && cb < 0) || (ca < 0 && cb > 0))) return true;
    return (ac == 0 && on_segment(c, a, b)) || (ad == 0 && on_segment(d, a, b)) ||
        (ca == 0 && on_segment(a, c, d)) || (cb == 0 && on_segment(b, c, d));
}
bool contains(const std::vector<float>& v, DoublePoint p) noexcept {
    if (v.size() < 6) return false;
    bool inside = false;
    for (std::size_t previous = v.size() - 2, current = 0; current < v.size(); previous = current, current += 2) {
        const auto a = point(v, previous), b = point(v, current);
        if (on_segment(p, a, b)) return true;
        if ((a.y > p.y) != (b.y > p.y) && p.x < (a.x - b.x) * (p.y - b.y) / (a.y - b.y) + b.x) inside = !inside;
    }
    return inside;
}
bool segment(const std::vector<float>& v, DoublePoint start, DoublePoint end) noexcept {
    if (v.size() < 6) return false;
    if (contains(v, start) || contains(v, end)) return true;
    for (std::size_t previous = v.size() - 2, current = 0; current < v.size(); previous = current, current += 2)
        if (crosses(start, end, point(v, previous), point(v, current))) return true;
    return false;
}
}
bool bounds_polygon_contains(const std::vector<float>& v, Point p) noexcept { return contains(v, {p.x, p.y}); }
bool bounds_polygon_segment(const std::vector<float>& v, Point start, Point end) noexcept { return segment(v, {start.x, start.y}, {end.x, end.y}); }
bool bounds_polygons_intersect(const std::vector<float>& a, const std::vector<float>& b) noexcept {
    if (a.size() < 6 || b.size() < 6) return false;
    if (contains(a, point(b, 0)) || contains(b, point(a, 0))) return true;
    for (std::size_t previous = a.size() - 2, current = 0; current < a.size(); previous = current, current += 2)
        if (segment(b, point(a, previous), point(a, current))) return true;
    return false;
}
bool bounds_aabb_segment(const BoundsAabb& aabb, Point start, Point end) noexcept {
    if (aabb.empty) return false;
    double low = 0, high = 1;
    const double dx = static_cast<double>(end.x) - start.x, dy = static_cast<double>(end.y) - start.y;
    const auto clip = [&](double p, double q) {
        if (p == 0) return q >= 0;
        const double t = q / p;
        if (p < 0) { if (t > high) return false; low = std::max(low, t); }
        else { if (t < low) return false; high = std::min(high, t); }
        return true;
    };
    return clip(-dx, static_cast<double>(start.x) - aabb.min_x) && clip(dx, static_cast<double>(aabb.max_x) - start.x) &&
        clip(-dy, static_cast<double>(start.y) - aabb.min_y) && clip(dy, static_cast<double>(aabb.max_y) - start.y);
}
}
